### PR TITLE
Add code to support domain management.

### PR DESCRIPTION
**Pull request by smneal**
### Context

This change facilitates managing the services domain and front door infrastructure.

- Ticket: https://trello.com/c/mNiwxEYG

### Changes proposed in this pull request

- Add Terraform code so the services domain can be managed from the repo.
- Add custom domains for dev and sandbox which don't have them.
- Add staging as a subdomain of the main domain.

### Guidance to review

- Run the following commands to:
1. Create the resource group: `make cpd_ecf deploy-domain-resources`
2. Create the DNS zone and front door `make cpd_ecf domains-infra-apply`
3. Deploy an environments records and front door configuration `make cpd_ecf dev domains-apply`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+.PHONY: sandbox
+sandbox:
+	$(eval DEPLOY_ENV=sandbox)
+
+.PHONY: production
+production:
+	$(eval DEPLOY_ENV=production)
+	$(if $(CONFIRM_PRODUCTION), , $(error Production can only run with CONFIRM_PRODUCTION))
+
+cpd_ecf:
```
Comment (saliceti) [inline Makefile:22]: Could this be renamed to something like `load_domain_config:` for clarity ?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+domain-azure-resources: set-azure-account set-azure-template-tag set-azure-resource-group-tags
+	az deployment sub create -l "UK South" --template-uri "https://raw.githubusercontent.com/DFE-Digital/tra-shared-services/${ARM_TEMPLATE_TAG}/azure/resourcedeploy.json" \
+		--name "${DNS_ZONE}domains-$(shell date +%Y%m%d%H%M%S)" --parameters "resourceGroupName=${RESOURCE_PREFIX}-${DNS_ZONE}domains-rg" 'tags=${RG_TAGS}' \
+			"tfStorageAccountName=${RESOURCE_PREFIX}${DNS_ZONE}domainstf" "tfStorageContainerName=${DNS_ZONE}domains-tf"  "keyVaultName=${RESOURCE_PREFIX}-${DNS_ZONE}domains-kv" ${WHAT_IF}
+
+validate-domain-resources: set-what-if domain-azure-resources # make cpd_ecf validate-domain-resources AUTO_APPROVE=1
+
+deploy-domain-resources: check-auto-approve domain-azure-resources # make cpd_ecf deploy-domain-resources AUTO_APPROVE=1
+
+domains-infra-init: set-production-subscription set-azure-account
```
Comment (saliceti) [inline Makefile:53]: cpd_ecf (or "load_domain_config") can be called here so there is no need to call it on command line

File: terraform/custom_domains/environment_domains/main.tf [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  environment           = each.value.environment_short
+  host_name             = each.value.origin_hostname
+  multiple_hosted_zones = var.multiple_hosted_zones
+  null_host_header      = try(each.value.null_host_header, false)
+  cached_paths          = try(each.value.cached_paths, [])
+}
+
+# Takes values from hosted_zone.domain_name.cnames (or txt_records, a-records). Use for domains which are not associated with front door.
+module "dns_records" {
+  source      = "git::https://github.com/DFE-Digital/terraform-modules.git//dns/records?ref=0.5.3"
```
Comment (saliceti) [inline terraform/custom_domains/environment_domains/main.tf:18]: Does it work with the `main` tag?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+load_domain_config:
+	$(eval include global_config/cpd_ecf_domain.sh)
+
+set-azure-account:
+	echo "Logging on to ${AZ_SUBSCRIPTION}"
+	az account set -s $(AZ_SUBSCRIPTION)
+
+set-azure-resource-group-tags: ##Tags that will be added to resource group on its creation in ARM template
+	$(eval RG_TAGS=$(shell echo '{"Portfolio": "Early Years and Schools Group", "Parent Business":"Teacher Training and Qualifications", "Product" : "Early Careers Framework", "Service Line": "Teaching Workforce", "Service": "Teacher services", "Service Offering": "Manage training for early career teachers", "Environment" : "$(ENV_TAG)"}' | jq . ))
```
Comment (peteryates) [inline Makefile:30]: Should there be an indication in here that ECF is part of CPD?

Comment (saliceti) [reply Makefile:30]: Matt is confirming the tags, we can amend later

Comment (smneal) [reply Makefile]: Updated Porfolio to `Teacher Continuing Professional Development`

**Review by saliceti — APPROVED** [review]

**Comment by smneal** [comment]
[ECF: Create DNS zone in Azure DNS via terraform and migrate](https://trello.com/c/mNiwxEYG/308-ecf-create-dns-zone-in-azure-dns-via-terraform-and-migrate)

**Review by peteryates — APPROVED** [review]
Nice, all looks good to me

**Pull request merged**